### ISSUE TITLE
[Packaging request] Please make Pandoc available as an AppImage / Integrate with Travis

### ISSUE THREAD

**Issue by KurtPfeifle** [jgm/pandoc#4029]
I want to request an automatically built AppImage package for Pandoc.

Actually, AppImages aren't "packages" in the the sense of other software packages. They bundle everything that is required into one file (which to the outside world appears as an ELF binary, but inside it is mostly an ISO file system tree) which is able to auto-mount itself to a mount point in */tmp/* and then run the application from there.

The most basic principle of the AppImage format is: `"1 file == 1 appliation"`. Contrary to what the first reading of their website may lead one to believe (*"this is for Desktop / GUI applications only"*), AppImages work very well for CLI tools too.**[^1]** <sub>*(I've successfully tested AppImages made from the `ippsample` software, provided by the ISTO Printer Working Group [PWG] for prototyping and testing IPP Everywhere [Internet Printing Protocol] software.)*</sub>.

The `"1 file == 1 application"` makes it very easy for users to make use of the software:

1. Download + save 1 file only.
1. Make it executable.
1. Run it (from whereever you stored it).

The `"1 file == 1 application"` makes it also very easy for developers to provide their software to their audience:

1. An AppImage runs on most Linux distribution without change.
1. Different versions of a software may be used side-by-side (important for testers / developers / reviewers / trainers).
1. Even officially un-released versions of the software can be easily made available to those who love to run (or have to test) the "bleeding edge".

It would be easy to integrate the automated build of an AppImage via Travis upon every single Git commit.

I'm pretty sure that @probono, the lead AppImage developer will help to setup an automatic workflow for Pandoc (maybe even prepare a pull request that you only need to test + accept). <sub>(After all, even Microsoft now builds [AppImages for their PowerShell 6.0.0 Betas](https://github.com/PowerShell/PowerShell/releases), capable to run on multiple Linux distributions.)</sub>

----

**[^1]:** One special case would apply to Pandoc: since it is only fully useful, if you also have the *`pandoc-citeproc`*-filter available, they should be bundled into the same AppImage. By creating two different symlinks some where in `$PATH` (f.e. `${HOME}/bin/`), *`pandoc`*  and *`pandoc-citeproc`* both pointing to the same executable AppImage, this both commands could be used independently from each other.

**Comment by mb21** [comment]
see also https://github.com/jgm/pandoc/issues/3399...

**Comment by jgm** [comment]
The binaries we provide in the binary packages are compiled with all required data files inside them.  So, *already* all you need is the binary: you can put this anywhere you like and run it.  What would AppImage add to this?

**Comment by KurtPfeifle** [comment]
> *"So, already all you need is the binary: you can put this anywhere you like and run it. What would AppImage add to this?"*

The "binary" is a *\*.tar.gz* or a *\*.deb* or an *\*.rpm*. That means, a user has to *install* it, usually with the help of a package manager (if *\*.rpm* or *\*.deb*). This will distribute the files around (according to the guidelines of the system's file system hierarchy requirements). This is fine -- this is how it all works for decades on Linux. (And there are even means to allow a user to use the *\*.tar.gz* version if he doesn't have root privileges -- but these are not for the faint-of-heart and not for less-technical users....)

An AppImage-d application you do not need to *install* in the classical sense. Just make this one file executable and run it, from anywhere you want (even from a USB drive). (Of course, once you start to use dozens of AppImages, you need to organize yourself a bit, and put them maybe all into one place). And you can run different versions of the same app side-by-side without any conflict. With the  *\*.tar.gz* or  *\*.deb* or *\*.rpm* approach you need to un-install or overwrite the old version (or let the package manager handle that), before you can run the new one. Once you run the new one, the old is gone for now. Much more hassle.

Also, I didn't mention: with the new update mechanism in place, updating AppImages is "cheap": because the updater downloads only a "diff" of the bytes which have changed (not a full new package, as with *\*.tar.gz* /  *\*.deb* / *\*.rpm*).

With this feature request I also have in mind integration with Travis and continuous builds. Right now, it is a big problem for me to get my hands on un-released versions. My local builds do not always complete (and I'm not a good enough technical person to solve the involved Haskell/stack/cabal problems). If Pandoc offered an automatically built AppImage triggered by every git commit (or once a day), I could use the same package on different (Linux) systems without much problem and spending much less time on the different build processes.

**Comment by probonopd** [comment]
I had made an experimental version a year ago:
https://bintray.com/probono/AppImages/Pandoc/1.17.2.glibc2.11#files

**Comment by jgm** [comment]
@probonopd - I get the impression that the AppImage includes not just the pandoc binary but any shared libraries it requires?  Is that correct?  If so, that's potentially an advantage over just distributing the pandoc binary by itself.  However, in the last release, we provide a fully static binary, which doesn't depend on any shared libraries.  So I'm still not seeing the advantage of an AppImage over just the bare binary.  Though it might get around one problem with the fully static binaries: they don't allow you to use lua filters that load lua modules written in C (`.so`).

Is @KurtPfeifle right that two different binaries could be included in the same AppImage and used via symlinks?

Also, what about documentation, copyright notices and the like?  This should always be distributed with the binary. Is it somehow included in the AppImage, and how?

@KurtPfeifle note that the pandoc-nightly repository has for a long time offered nightly binary builds which you can use in CI.

**Comment by KurtPfeifle** [comment]
> *"I had made an experimental version a year ago:
https://bintray.com/probono/AppImages/Pandoc/1.17.2.glibc2.11#files"*

That's nice -- I didn't notice at the time. I guess it does not include *`pandoc-citeproc`* yet? And the symlink-/ARGV[0]-feature is not implemented?

A new, updated expermental version for Pandoc-2.0.1 would be nice. (Honestly, nobody  -- including me --  will be interested in v1.17.2 any more...)

**Comment by KurtPfeifle** [comment]
> *"@KurtPfeifle note that the pandoc-nightly repository has for a long time offered nightly binary builds which you can use in CI."*

Thanks for this info. I wasn't aware of that. Will give it a spin.

**Comment by probonopd** [comment]
> That's nice -- I didn't notice at the time. I guess it does not include pandoc-citeproc yet? And the symlink-/ARGV[0]-feature is not implemented?

No, I implemented that feature only this week when you requested it ;-)

> A new, updated expermental version for Pandoc-2.0.1 would be nice.

If there is interest then I can probably make a Pull Request that would build one on Travis CI on each `git push`. But before I do this I would like to understand the level of interest from the project.

**Comment by KurtPfeifle** [comment]
> *"But before I do this I would like to understand the level of interest from the project."*

For the project itself, only John and other contributors can speak.

From the point of view of users who want to have a continuous look at new features as they develop, while at the same time keeping their "stable" installations in place for "production", my level of interest cannot be over-estimated. :-)

And this is true for all the software which I regularly (must) use for work- and hobby-related things. These are mostly CLI tools (not necessarily the main target for AppImage building): CUPS, cups-filters, Ghostscript/GhostPDL, MuPDF, ImageMagick, Poppler-tools, TabulaPDF, QPDF, IPPSample, NeoVim and Pandoc. I would be in software heaven if I had easy access to the latest as well as to different past releases for all these. Sometimes, to solve a support question concerning an older version of a software, it would be so much easier to just start the respective AppImage instead of all the other fuddling which is needed (sometimes including complete VMs).

I'm pretty sure a well-working (and automatically CI-built) Pandoc AppImage would quickly find its fan community. But be warned: these fans will then start asking about more AppImages for all their other pet software. If you do this work, it will create more work for you :-)

**Comment by jgm** [comment]
You can already get up-to-date builds of the dev version from pandoc-nightly.  But I'm very reluctant to have something that looks like an official release and is produced on a rolling basis. Random commits are often broken; we have releases for a reason. pandoc-nightly is very clearly marked as being just for testing and confirming bugs.

**Comment by probonopd** [comment]
@jgm pandoc-nighly.AppImage would be as official or inofficial as you like it to be, not different from the same software in a .tar.gz or any other archive format. The only difference really would be that in addition to being able to extract the contents of an AppImage (like any other archive), you can also directly execute it (which will make it mount itself and execute whatever you put inside).

**Comment by KurtPfeifle** [comment]
> *"Random commits are often broken; we have releases for a reason."*

I'm aware of that...

> *"pandoc-nightly is very clearly marked as being just for testing and confirming bugs."*

That's exactly what I want to use them for.

> *"But I'm very reluctant to have something that looks like an official release and is produced on a rolling basis."*

I understand that...

**Comment by jgm** [comment]
> >"pandoc-nightly is very clearly marked as being just for testing and confirming bugs."
>
> That's exactly what I want to use them for.

If that is your intended use, I'm not really seeing what AppImage buys us over what we already have.  You can already get the binaries you need for testing quite easily from pandoc-nightly.  I can see how AppImage would be very useful for apps with complex dependencies on shared libraries or external data files.  But that isn't the case here.

**Comment by KurtPfeifle** [comment]
>>> *"pandoc-nightly is very clearly marked as being just for testing and confirming bugs."*
>>>
>> *"That's exactly what I want to use them for."*
>>
> *"If that is your intended use, I'm not really seeing what AppImage buys us over what we already have."*

It's not my ONLY intended use case. Other use cases were mentioned.

But even for that one use case: using the pandoc-nightly will STILL require me to "manage" the different installable versions, and install, run, de-install/replace them. I cannot (easily) do that side-by side.  With AppImages the "management" side of things is just to name them according to their version and start them. It's on the level of *file management*. And I can carry them along on a USB stick to every (Linux) machine I use and directly start them from there. Can I do this with pandoc-nightly?

I'll use pandoc-nightly for now. But I keep dreaming of automatically generated *`pandoc-nightly.AppImage`*.... :-)

**Comment by mb21** [comment]
From what I'm hearing here, @jgm is reluctant to add another officially supported package format (and the overhead this would mean for him). But it sounds like a great fit for pandoc-nightly, why not ask whether pull requests are welcome there?

**Comment by KurtPfeifle** [comment]
> *" [...] sounds like a great fit for pandoc-nightly, why not ask whether pull requests are welcome there?"*

Good idea, @mb21, thanks! I've to find out who owns pandoc-nightly then. From @jgm's comments I had (wrongly) taken the impression that it's officially part of the overall Pandoc project.

**Comment by mb21** [comment]
I'd open an issue on https://github.com/pandoc-extras/pandoc-nightly, or the people behind are also around on pandoc-discuss...

**Comment by jgm** [comment]
+++ Kurt Pfeifle [Nov 03 17 08:07 ]:
>   It's not my ONLY intended use case. Other use cases were mentioned.

So, can you say what use case isn't handled by
pandoc-nightly, and also isn't going to encourage people to
run half-baked nightlies for purposes other than testing?

>   But even for that one use case: using the pandoc-nightly will STILL
>   require me to "manage" the different installable versions, and install,
>   run, de-install/replace them. I cannot (easily) do that side-by side.
>   With AppImages the "management" side of things is just to name them
>   according to their version and start them. It's on the level of file
>   management.

It's the same with pandoc-nightly.  You get a zipped
executable.  Download it, rename it, put it where you like
and run it.  The only difference I see between that an an
AppImage is that you don't have to unzip the AppImage. But
that's pretty minor.

>   And I can carry them along on a USB stick to every (Linux)
>   machine I use and directly start them from there. Can I do this with
>   pandoc-nightly?

Probably not, because as far as I know, they're not using
the new fully static build configuration.  So there will be
some dynamic library dependencies (which will be satisfied
on most linux systems).  But that could change if
they used the new static build configuration.

**Comment by jgm** [comment]
@probonopd - I still have two questions about AppImage:

1.  Do the AppImage containers contain documentation, copyright notices, man pages, etc.?
2. Pandoc is usually distributed with two binaries, pandoc and pandoc-citeproc.  Could these be put in the same AppImage container, and handled using symlinks, as @KurtPfeifle suggests?

**Comment by probonopd** [comment]
> Do the AppImage containers contain documentation, copyright notices, man pages, etc.?

Yes, many do, in the usual `usr/share/.../` inside the AppImage. But ultimately it is up to the person who puts together the AppImage to decide what goes into it. 

> Pandoc is usually distributed with two binaries, pandoc and pandoc-citeproc. Could these be put in the same AppImage container, and handled using symlinks, as @KurtPfeifle suggests?

Yes.

**Comment by jgm** [comment]
+++ probonopd [Nov 03 17 17:01 ]:
>   Yes, many do, in the usual usr/share/.../ inside the AppImage. But
>   ultimately it is up to the person who puts together the AppImage to
>   decide what goes into it.

How does the user of the AppImage access these things?
(Or do they just have to manually unpack it?)

**Comment by probonopd** [comment]
When the user executes the AppImage, the contents of the filesystem image are temporarily mounted, and the payload application is executed from the mounted filesystem. The payload application can access these things from paths relative to itself. After the payload application exits, the filesystem image is unmounted.

**Comment by svenevs** [comment]
> But even for that one use case: using the pandoc-nightly will STILL require me to "manage" the different installable versions, and install, run, de-install/replace them. I cannot (easily) do that side-by side.

@KurtPfeifle the typical development technique to handle switching between multiple different versions of software stacks is through "environment modules".  **Warning**: if you have never used them, understanding how to get them working can be a challenge.  You may also be interested in [lmod](https://www.tacc.utexas.edu/research-development/tacc-projects/lmod), but that's overkill for just managing multiple pandoc.  I mention it because if you are just getting started with environment modules, and you have a use for `lmod`, it may be better to just go for gold :)

I'm not commenting on AppImage component.  Based on some of what has been written here, it sounds like you may not be using environment modules.  Using them may make your life much easier when managing multiple installations.  For example, you can update to the nightly build for pandoc at your leisure, call this module `pandoc-nightly-2017-12-21` (and keep as many versions as storage permits I suppose, or just keep one or two around).  But at any point if you need to use the stable, just load that module instead.  All you really need your module to do is change the `$PATH` variable and you're done.  I hope that helps some.

**Comment by probonopd** [comment]
While I am by no means a Pandoc expert I doubt many users know how to use environment modules. In fact, as a _user_ I don't even want to be bothered to learn what programming language/framework Pandoc is written in. As a _user_, I simply want to download a single file, and use that, without it interfering with the rest of the system, such as any other versions that may or may not already be available there. Just my 2c though.

**Comment by ickc** [comment]
Hi, pandoc-nightly maintainer here. Sorry I completely missed this discussion as I found that if I’m not in a team in a GitHub Organization, I won’t get notification (as a similar issue opened there) even if I’m the owner.

A few comments:

- I have no knowledge on AppImage, and from my background it’s unlikely I’m going to spend time learning it. However, I can either accept pull request to pandoc-nightly to bundle them together, or I can give you access to pandoc-extras organization and add a repo (say, pandoc-appimage for released, versioned pandoc, pandoc-nightly-appimage for nightly). If you give me a script to build it (using Ubuntu), I can help setting up cron job to auto-build daily nightly builds.

- I agree that pandoc-nightly can in principle get the job done. I understand the complication about having 2 executables, but except for that it is self-contained, and 2 is a managible number. I heard from above that there’s some convenience benefits from AppImage when switching between versions (although iI don’t exactly know the details), I imagine I can quickly setup something to allow me to rapidly changing between different versions using some standard shell tricks. e.g. right now the quickest way to set it up comes to my mind is this:

    1. get a list of versions / commits I want to test
    2. write a script to wget/curl them all using xargs into first level subdir.
    3. write a make file, having a target `v%` which is to be the version/commit, e.g. when `make v123456`, `mkdir -p pandoc; ln -s pandoc-$*/pandoc pandoc/pandoc ...`
    4. `export PATH=“./pandoc:$PATH”`

    Done. This makefile can help you context switch instantly. By the way, this is exactly those situations pandoc-nightly and pandoc-portable will be useful: getting a bunch of them running side by side becomes easy.

Regardless if existing tool can get the job done or not, I think options are good under the umbrella of pandoc-extras, as long as someone is going to maintain it ;)

**Comment by ickc** [comment]
>It would be easy to integrate the automated build of an AppImage via Travis upon every single Git commit.

Note that this is going to be expensive. Pandoc’s current Travis build turns off many optimizations, so that e.g. the final binary will run slow and is larger, AFAIK. Pandoc-nightly instead build it as usual, so e.g. the binary you get from an official release should be the same as the pandoc nightly at that commit. These can be important in testing (e.g. you don’t want to keep multiple copies of large binaries, or you want to see the performance accross commits). If you want to change this in the official pandoc’s travis builds, then it will becomes much slower. And from my expereince, @jgm’s Travis is already pretty busy and slow. So this is the adavantage of seperating Ci snd nightly build. And one day each should be fine-grain enough. Are there any particular reason you’d want to have a binary per commit?

**Comment by ickc** [comment]
Since environment modules has been mentioned, just to add some comments on this:

1. a quick wiki seems to suggest that it is popular amount computer centers, notably it mentioned NERSC managed almost all softwares using environment modules. This is no longer true though. The 2 other popular tools to setup environments on NERSC are docker/shifter in general and conda for Python stack (by the way, one can setup conda env with different versions of pandoc too).

2. module is not difficult to use, but difficult to setup. It can be convenient for users, but since the intended use here seems to be testing, it is overkill.

3. one can imagine one can also provide “modulefiles templates” for pandoc to ease setting it up. So it can easily be seen that options like docker, AppImage, module, etc. can be overwhelming to be provided in official pandoc. (But they are welcome in pandoc-extras!)
